src/codec.rs

```diff
@@ -207,31 +227,43 @@ impl<T> HasCompact for T where
 
 // Note: we use *LOW BITS* of the LSB in LE encoding to encode the 2 bit key.
 
-impl Encode for Compact<u8> {
+impl<'a> Encode for CompactRef<'a, u8> {
 	fn encode_to<W: Output>(&self, dest: &mut W) {
 		match self.0 {
 			0...0b00111111 => dest.push_byte(self.0 << 2),
-			_ => (((self.0 as u16) << 2) | 0b01).encode_to(dest),
+			_ => (((*self.0 as u16) << 2) | 0b01).encode_to(dest),
 		}
 	}
 }
 
-impl Encode for Compact<u16> {
+impl Encode for Compact<u8> {
+	fn encode_to<W: Output>(&self, dest: &mut W) {
+		CompactRef(&self.0).encode_to(dest)
+	}
+}
+
+impl<'a> Encode for CompactRef<'a, u16> {
 	fn encode_to<W: Output>(&self, dest: &mut W) {
 		match self.0 {
-			0...0b00111111 => dest.push_byte((self.0 as u8) << 2),
-			0...0b00111111_11111111 => ((self.0 << 2) | 0b01).encode_to(dest),
-			_ => (((self.0 as u32) << 2) | 0b10).encode_to(dest),
+			0...0b00111111 => dest.push_byte((*self.0 as u8) << 2),
```

Comment:
It would be very nice to have some more comments on code like this.
However, I can understand that this might be out of range for this PR.
Not stopping merging because of this.

Reply:
There is already documentation for it in the source file.